New resource: `azurerm_container_registry_task`

magodo opened the pull request:
This PR adds support for a new resource `azurerm_container_registry_task`.

To run the tests you need to specify following environment variables:

- `ARM_TEST_ACR_TASK_GITHUB_REPO_URL`: URL to a forked repo from the https://github.com/Azure-Samples/acr-build-helloworld-node
- `ARM_TEST_ACR_TASK_GITHUB_USER_TOKEN`: Github personal access token with at least the `repo:status` and `public_repo` permissions.

See https://docs.microsoft.com/en-us/azure/container-registry/container-registry-tutorial-build-task for details.

Fixes #9110

## Test

```shell
💢 TF_ACC=1 go test -v -timeout=20h ./internal/services/containers -run='TestAccContainerRegistryTask_'
=== RUN   TestAccContainerRegistryTask_dockerStep
=== PAUSE TestAccContainerRegistryTask_dockerStep
=== RUN   TestAccContainerRegistryTask_fileTaskStep
=== PAUSE TestAccContainerRegistryTask_fileTaskStep
=== RUN   TestAccContainerRegistryTask_encodedTaskStep
=== PAUSE TestAccContainerRegistryTask_encodedTaskStep
=== RUN   TestAccContainerRegistryTask_dockerStepBaseImageTrigger
=== PAUSE TestAccContainerRegistryTask_dockerStepBaseImageTrigger
=== RUN   TestAccContainerRegistryTask_dockerStepSourceTrigger
=== PAUSE TestAccContainerRegistryTask_dockerStepSourceTrigger
=== RUN   TestAccContainerRegistryTask_dockerStepTimerTrigger
=== PAUSE TestAccContainerRegistryTask_dockerStepTimerTrigger
=== RUN   TestAccContainerRegistryTask_identity
=== PAUSE TestAccContainerRegistryTask_identity
=== RUN   TestAccContainerRegistryTask_fileTaskStepRegistryCredential
=== PAUSE TestAccContainerRegistryTask_fileTaskStepRegistryCredential
=== RUN   TestAccContainerRegistryTask_systemTask
=== PAUSE TestAccContainerRegistryTask_systemTask
=== RUN   TestAccContainerRegistryTask_requiresImport
=== PAUSE TestAccContainerRegistryTask_requiresImport
=== CONT  TestAccContainerRegistryTask_dockerStep
=== CONT  TestAccContainerRegistryTask_dockerStepTimerTrigger
=== CONT  TestAccContainerRegistryTask_encodedTaskStep
=== CONT  TestAccContainerRegistryTask_systemTask
=== CONT  TestAccContainerRegistryTask_dockerStepBaseImageTrigger
=== CONT  TestAccContainerRegistryTask_dockerStepSourceTrigger
=== CONT  TestAccContainerRegistryTask_fileTaskStep
=== CONT  TestAccContainerRegistryTask_requiresImport
--- PASS: TestAccContainerRegistryTask_systemTask (163.74s)
=== CONT  TestAccContainerRegistryTask_identity
--- PASS: TestAccContainerRegistryTask_encodedTaskStep (164.41s)
=== CONT  TestAccContainerRegistryTask_fileTaskStepRegistryCredential
--- PASS: TestAccContainerRegistryTask_dockerStepTimerTrigger (168.60s)
--- PASS: TestAccContainerRegistryTask_dockerStepSourceTrigger (168.67s)
--- PASS: TestAccContainerRegistryTask_requiresImport (173.38s)
--- PASS: TestAccContainerRegistryTask_dockerStep (181.19s)
--- PASS: TestAccContainerRegistryTask_dockerStepBaseImageTrigger (229.51s)
--- PASS: TestAccContainerRegistryTask_fileTaskStep (230.03s)
--- PASS: TestAccContainerRegistryTask_fileTaskStepRegistryCredential (239.12s)
--- PASS: TestAccContainerRegistryTask_identity (261.89s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/containers    425.663s
```

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+			ForceNew:     true,
+			ValidateFunc: validate.RegistryID,
+		},
+		"platform_setting": {
```
settings seems redundant?
```suggestion
		"platform": {
```

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+				},
+			},
+		},
+		"docker_step": {
```
step seems redundant given there can only be one? ie this is a task.docker.ect

katbyte replied:
```suggestion
		"docker": {
```

magodo replied:
I prefer to keep the "step" here and the same for the other two types of steps, so to be more explicit that this block is about the configurations of the steps. Otherwise, the "docker", "file", "encoded" are not so obvious that they are related to the step, which might cause users to take some time to figure them out. WDYT?

katbyte replied:
👍

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+							Type: pluginsdk.TypeString,
+						},
+					},
+					"is_push_enabled": {
```
we don't use is in boolean properties?
```suggestion
					"push_enabled": {
```

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+						Optional: true,
+						Default:  true,
+					},
+					"is_cache_enabled": {
```
```suggestion
					"cache_enabled": {
```

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+			},
+			ConflictsWith: []string{"file_task_step", "encoded_task_step"},
+		},
+		"file_task_step": {
```
```suggestion
		"file": {
```

task is in the resource name

katbyte replied:
going off your comment above: 
```suggestion
		"file_step": {
```

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*schema.Schema{
+					"task_file_path": {
```
task and file are already in the resource and block names?
```suggestion
					"path": {
```

magodo replied:
Shall we be more explict that this *path* here is the task file's path, rather than some other path (e.g. value file path, context path, etc.)?

magodo replied:
I still feel `task_file_path` is better with the presence of `value_file_path` in the same block.

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+						Required:     true,
+						ValidateFunc: validation.StringIsNotEmpty,
+					},
+					"value_file_path": {
```
```suggestion
					"value_path": {
```

magodo replied:
If we keep the `task_file_path`, then the `value_file_path` looks more consistent?

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+			},
+			ConflictsWith: []string{"docker_step", "encoded_task_step"},
+		},
+		"encoded_task_step": {
```
```suggestion
		"encoded": {
```

katbyte replied:
```suggestion
		"encoded_step": {
```
ect

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+			MaxItems: 1,
+			Elem: &pluginsdk.Resource{
+				Schema: map[string]*schema.Schema{
+					"task_content": {
```
```suggestion
					"content": {
```

magodo replied:
Shall we be more explicit here that the `content` is `task_content` rather than the `value_content`?

katbyte replied:
sure

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+							}, false),
+						},
+					},
+					"source_setting": {
```
this block is redundant? all thes properties oculd be moved up a level with no consequence?

katbyte commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1573 @@
+									Type:     pluginsdk.TypeString,
+									Optional: true,
+								},
+								"auth": {
```
could we use `authentitication`?

magodo commented:
@katbyte Thank you for the initial review. I've made some of the renaming changes and the flattening schema change. While there are some properties I still feel it is more appropriate to keep them as it was so to be more explicit and easy to discover?

magodo commented:
@katbyte I've renamed the `encoded_step` and `file_step`, while I still prefer to using the `task_file_path` inside the `file_step` block as it makes more sense when co-exist with the `value_file_path`. WDYT?

koikonom commented on internal/services/containers/container_registry_task_resource.go:
```diff
@@ -0,0 +1,1551 @@
+	}
+
+	out := legacyacr.Credentials{
+		SourceRegistry:   &legacyacr.SourceRegistryCredentials{LoginMode: legacyacr.SourceRegistryLoginModeDefault},
```
I see in the API docs that there's also another login mode called "None" (I'm assuming it means it should not try to autenticate at all?). 

Are we skipping it on purpose?

koikonom reviewed (changes requested):
Hi @magodo, thanks for opening this PR! I only have a couple of things I'd love to hear what you think.

magodo commented:
@koikonom Thank you for the good suggestions! I've updated the `Update` logic to using the `HasChanged()`, also modified the `registry_credential` block to involve the `login_mode` property. Please take another look!

koikonom commented:
LGTM!
